# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/loader/ModuleLoadRequest.cpp
# Commit: e47abb2b8ac7
# Full Hash: e47abb2b8ac70efe89371c67766def82374c0f36
# Author: Alexandru Marc <amarc@mozilla.com>
# Date: 2025-09-27 12:29 +0300
# Regressor Bug: 1968890
# File Overlap Count: 14
# Overlapping Files: dom/base/test/jsmodules/chrome.toml, dom/base/test/jsmodules/test_dynamicImportErrorMessage.html, dom/script/ScriptLoadContext.cpp, dom/script/ModuleLoader.cpp, dom/base/test/jsmodules/import_404.mjs
# Description:
#   Revert "Bug 1968890 - Part 9: Add test for dynamic import. r=jonco" as potential cause of javascript engine crashes (bug 1991268)
# ==============================================================================

diff -r 7e2611c97326 -r e47abb2b8ac7 js/loader/ModuleLoadRequest.cpp
--- a/js/loader/ModuleLoadRequest.cpp	Sat Sep 27 11:41:54 2025 +0300
+++ b/js/loader/ModuleLoadRequest.cpp	Sat Sep 27 12:29:56 2025 +0300
@@ -53,9 +53,9 @@
     ModuleLoadRequest* aRootModule)
     : ScriptLoadRequest(ScriptKind::eModule, aURI, aReferrerPolicy,
                         aFetchOptions, aIntegrity, aReferrer, aContext),
-      mKind(aKind),
+      mIsTopLevel(aKind == Kind::TopLevel || aKind == Kind::DynamicImport),
       mModuleType(aModuleType),
-      mErroredLoadingImports(false),
+      mIsDynamicImport(aKind == Kind::DynamicImport),
       mLoader(aLoader),
       mRootModule(aRootModule) {
   MOZ_ASSERT(mLoader);
@@ -66,12 +66,23 @@
 }
 
 bool ModuleLoadRequest::IsErrored() const {
-  if (!mModuleScript || mErroredLoadingImports) {
-    return true;
+  return !mModuleScript || mModuleScript->HasParseError();
+}
+
+void ModuleLoadRequest::Cancel() {
+  if (IsCanceled()) {
+    return;
   }
 
-  MOZ_ASSERT_IF(mModuleScript->HasErrorToRethrow(), !IsDynamicImport());
-  return mModuleScript->HasParseError() || mModuleScript->HasErrorToRethrow();
+  if (IsFinished()) {
+    return;
+  }
+
+  ScriptLoadRequest::Cancel();
+
+  mModuleScript = nullptr;
+  mReferrerScript = nullptr;
+  mModuleRequestObj = nullptr;
 }
 
 void ModuleLoadRequest::SetReady() {
@@ -97,6 +108,10 @@
   MOZ_ASSERT(IsFetching() || IsPendingFetchingError());
 
   mModuleScript = mLoader->GetFetchedModule(ModuleMapKey(mURI, mModuleType));
+  if (IsErrored()) {
+    ModuleErrored();
+    return;
+  }
 }
 
 void ModuleLoadRequest::LoadFailed() {
@@ -126,7 +141,13 @@
   }
 
   MOZ_ASSERT(!IsFinished());
-  MOZ_ASSERT(IsErrored());
+
+  mozilla::DebugOnly<bool> hasRethrow =
+      mModuleScript && mModuleScript->HasErrorToRethrow();
+
+  // When LoadRequestedModules fails, we will set error to rethrow to the module
+  // script and call ModuleErrored().
+  MOZ_ASSERT(IsErrored() || hasRethrow);
 
   if (IsFinished()) {
     // Cancelling an outstanding import will error this request.
@@ -139,7 +160,7 @@
 
 void ModuleLoadRequest::LoadFinished() {
   RefPtr<ModuleLoadRequest> request(this);
-  if (IsDynamicImport()) {
+  if (IsTopLevel() && IsDynamicImport()) {
     mLoader->RemoveDynamicImport(request);
   }
 